### PR TITLE
Migrate new reaction job 5305

### PR DESCRIPTION
**Pull request by mukesh4139**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [X] Optimization
- [ ] Documentation Update

## Description
This change creates a Sidekiq Worker equivalent of Notifications::NewReactionJob

## Related Tickets & Documents
#5305 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

PS: This is my second project for Open Source Contribution. And the first project which is on production. Excited :)

### CODE REVIEW

File: spec/models/notification_spec.rb [hunk 1]
```diff
@@ -272,24 +272,22 @@
     end
```
Comment (rhymes) [inline spec/models/notification_spec.rb]: I noticed how the logic of the specs has become more complicated, why did you take this approach?

File: spec/models/notification_spec.rb [hunk 2]
```diff
@@ -272,24 +272,22 @@
...
-        reaction = create(:reaction, reactable: comment, user: user)
+      let(:reaction) { create(:reaction, reactable: comment, user: user) }
+      let(:count) { comment.user.notifications.count }
 
-        perform_enqueued_jobs do
-          expect do
-            described_class.send_reaction_notification(reaction, reaction.reactable.user)
-          end.to change(comment.user.notifications, :count).by(1)
+      before do
+        @previous_notification_count = comment.user.notifications.count
```
Comment (rhymes) [inline spec/models/notification_spec.rb]: we shouldn't use instance variables, did you have any issues keeping the style of the specs as is? let me know

File: spec/models/notification_spec.rb [hunk 3]
```diff
@@ -298,6 +296,18 @@
       let!(:notification) { create(:notification, user: user, notifiable: comment, action: "Reaction") }
       let(:sibling_reaction) { create(:reaction, reactable: comment, user: user3) }
 
+      before do
```
Comment (rhymes) [inline spec/models/notification_spec.rb]: why this?

File: app/workers/notifications/new_reaction_worker.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module Notifications
+  class NewReactionWorker
+    include Sidekiq::Worker
+
+    sidekiq_options queue: :medium_priority, retry: 10
+
+    def perform(reaction_data, receiver_data, service = Notifications::Reactions::Send)
```
Comment (rhymes) [inline app/workers/notifications/new_reaction_worker.rb]: you can remove the service as a parameter and mock `Notifications::Reactions::Send` in the worker specs directly

File: spec/workers/notifications/new_reaction_worker_spec.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      expect(reaction_service).to have_received(:call).with(reaction_data, org).once
+    end
+
+    it "doesn't call if is a receiver is of a wrong class" do
+      subject.perform(reaction_data, { klass: "Tag", id: 10 }, reaction_service)
+      expect(reaction_service).not_to have_received(:call)
+    end
+
+    it "doesn't call if is a receiver doesn't exist" do
+      subject.perform(reaction_data, { klass: "Organization", id: org.id + 1 }, reaction_service)
```
Comment (rhymes) [inline spec/workers/notifications/new_reaction_worker_spec.rb]: keep in mind that id + 1 could technically exist, you can pass nil to make sure it doesn't exist

File: spec/models/notification_spec.rb (hunk 1 above)
Comment (mukesh4139) [reply spec/models/notification_spec.rb]: @rhymes thank you for your review comments.

`sidekiq_perform_enqueued_jobs` is only working in a `before` block. That's why I had to make things little complex. 

previously used expect{}.to change() is making sure that Sidekiq Jobs are executed.

For eg. this doesn't execute all enqueued Jobs before checking
```ruby
it "example" do
  sidekiq_perform_enqueued_jobs do 
    described_class.send_reaction_notification(reaction, reaction.reactable.user)
    expect {}.to change().by(1)
  end
end
```

If you can guide on this, I can make specs simple are previous.

Comment (rhymes) [reply spec/models/notification_spec.rb]: what if you do something like:

```ruby
expect do
  sidekiq_perform_enqueued_jobs do
    ... send reaction ...
  end
end.to change...

File: spec/models/notification_spec.rb (hunk 3 above)
Comment (mukesh4139) [reply spec/models/notification_spec.rb]: For the same reason mentioned here: https://github.com/thepracticaldev/dev.to/pull/5401#discussion_r364198021

File: spec/models/notification_spec.rb (hunk 1 above)
Comment (mukesh4139) [reply spec/models/notification_spec.rb]: Thank you @rhymes 
It worked like charm. I will update specs.

File: app/workers/notifications/new_reaction_worker.rb (hunk 4 above)
Comment (mukeshchdy) [reply app/workers/notifications/new_reaction_worker.rb]: I am sorry I didn't get you here.
I have already used `let(:reaction_service) { double }` to make fake reaction_service.

Comment (rhymes) [reply app/workers/notifications/new_reaction_worker.rb]: What I mean is that you don't need really need to pass the service as an argument and you can test the real one in the specs. I'llve leave you some suggestions

```suggestion
    def perform(reaction_data, receiver_data)
```

File: app/workers/notifications/new_reaction_worker.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+      # Sidekiq Parameters are hash with stringified keys, so we need to symbolize keys
+      receiver_data = receiver_data.symbolize_keys
+      reaction_data = reaction_data.symbolize_keys
+
+      receiver_klass = receiver_data.fetch(:klass)
+      return unless %w[User Organization].include?(receiver_klass)
+
+      receiver = receiver_klass.constantize.find_by(id: receiver_data.fetch(:id))
+      service.call(reaction_data, receiver) if receiver
```
Comment (rhymes) [inline app/workers/notifications/new_reaction_worker.rb]:
```suggestion
      Notifications::Reactions::Send.call(reaction_data, receiver) if receiver
```

File: spec/models/notification_spec.rb [hunk 7]
```diff
@@ -325,20 +325,22 @@
         expect(notification.json_data["user"]["id"]).to eq(user3.id)
       end
 
-      it "creates and destroys the notification properly" do
+      it "creates and destroys the notification properly", execute_create_and_destroy: true do
```
Comment (rhymes) [inline spec/models/notification_spec.rb]: `execute_create_and_destroy: true` is not needed here

File: spec/workers/notifications/new_reaction_worker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+RSpec.describe Notifications::NewReactionWorker, type: :worker do
+  let(:reaction_data) { { reactable_type: "Comment", reactable_id: 1, reactable_user_id: 2 } }
+  let(:org) { create(:organization) }
+  let(:receiver_data) { { klass: "Organization", id: org.id } }
+  let(:worker) { subject }
+
+  include_examples "#enqueues_on_correct_queue", "medium_priority", [{}, {}, true]
+
+  describe "#perform" do
+    let(:reaction_service) { double }
```
Comment (rhymes) [inline spec/workers/notifications/new_reaction_worker_spec.rb]:
```suggestion
    let(:reaction_service) { Notifications::Reactions::Send }
```

File: spec/workers/notifications/new_reaction_worker_spec.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+    before { allow(reaction_service).to receive(:call) }
+
+    it "calls the service" do
+      subject.perform(reaction_data, receiver_data, reaction_service)
+      expect(reaction_service).to have_received(:call).with(reaction_data, org).once
```
Comment (rhymes) [inline spec/workers/notifications/new_reaction_worker_spec.rb]: and in the test you do this:

```suggestion
      allow(reaction_service).to receive(:call).with(reaction_data, org)

      subject.perform(reaction_data, receiver_data)

      expect(reaction_service).to have_received(:call).with(reaction_data, org).once
```

same for the others below

**Comment by mukesh4139** [comment]
Hi @rhymes, thank you for your help. I have updated the specs

**Comment by rhymes** [comment]
Hi @mukesh4139 you modified the wrong file :) `spec/jobs/notifications/new_reaction_job_spec.rb` instead of `spec/workers/notifications/new_reaction_worker_spec.rb` :D 

Perfectly understandable hehe

**Comment by mukesh4139** [comment]
Hi @rhymes, yes I realized and fixed it.

**Pull request merged**